cgsmiles/sample.py

```diff
@@ -7,133 +7,240 @@
+        self.fragments_by_bonding = defaultdict(list)
+        if fragment_masses:
+            guess_mass_from_PTE = False
+            self.fragment_masses = fragment_masses
+        elif all_atom:
+            self.fragment_masses = {}
+            guess_mass_from_PTE = True
+        else:
+            msg = ("No fragment masses were provided but the resolution"
+                   "is not all_atom. We cannot guess masses for abitrary"
+                   "CG molecules.")
+            raise IOError(msg)
 
         # we need to store which bonding descriptors is present in which
         # fragment so we can later just look them up
-        self.fragments_by_bonding = defaultdict(list)
         for fragname, fraggraph  in self.fragment_dict.items():
+            if guess_mass_from_PTE:
+                mass = compute_mass(fraggraph)
+                self.fragment_masses[fragname] = mass
             bondings = nx.get_node_attributes(fraggraph, "bonding")
             for node, bondings in bondings.items():
                 for bonding in bondings:
                     self.fragments_by_bonding[bonding].append((fragname, node))
 
-    def update_open_bonds(self):
-        self.current_open_bonds = defaultdict(list)
-        open_bonds = nx.get_node_attributes(self.molecule, 'bonding')
-        for node, bonding_types in open_bonds.items():
-            for bonding_types in bonding_types:
-                self.current_open_bonds[bonding_types].append(node)
-
-    def pick_bonding(self):
+    def grow_chain(self, molecule, seed=None):
         """
-        Given a momomer to connect to the growing molecule see if there is
-        a matching bonding descriptor and what the probabilitiy of forming
-        this bond is.
+        Pick an open bonding descriptor according to `bonding_probabilities`
+        and then pick a fragment that has the complementory bonding descriptor.
+
+        Parameters
+        ----------
+        molecule: nx.Graph
+            the molecule to extend
+
+        Returns
+        -------
+        nx.Graph
+            the grown molecule
+        str
+            the fragment name of the added fragment
         """
-        probs = [self.bonding_probabilities[bond_type] for bond_type in self.current_open_bonds]
+        # 1. get the probabilties of any bonding descriptor on the chain to
+        #    form the new bond
+        probs = np.array([self.bonding_probabilities[bond_type[:-1]] for bond_type in self.current_open_bonds])
+        probs = probs / sum(probs)
+        # 2. pick a random bonding descriptor according to these probs
         bonding = np.random.choice(list(self.current_open_bonds.keys()), p=probs)
-        source_node = np.random.choice(self.current_open_bonds[bonding])
-        compl_bonding = _find_complementary_bonding_descriptor(bonding)
+        # 3. get a corresponding node; it may be that one descriptor is found on
+        #    several nodes
+        random.seed(a=seed)
```

Comment:
This needs to move to the `__init__`, because the descriptor is picked *before* this

Reply:
done